xxxx-ewallet-blockchain-integration.md

```diff
@@ -0,0 +1,339 @@
+- `create_eth_token` (publish a token smart contract on Ethereum)
+- `mint` (adds more value to an existing token on Ethereum)
+- `deposit` (deposit funds to the specified Plasma contract)
+- `exit` (exit funds from the specified Plasma contract)
+- `send_rootchain_transaction` (sends an Ethereum transaction)
+- `send_chain_transaction` (sends a Plasma transaction)
+- `get_quote` (get a quote from Exchange)
+- `place_exchange_order` (place an order for exchange)
+- `register_listener` (register a GenServer as a listener to events from Ethereum and Plasma)
+
+#### Database Changes
+
+Table name: `contract_templates` (New)
+
+- `uuid`
+- `name`
+- `content`
+- `inserted_at`
+- `updated_at`
+
+Table name: `blockchain_wallet` (New)
+
+- `uuid`
+- `address`
+- `pub_key`
+- `encrypted_priv_key` (only for `hot` wallets, encrypted using a key defined in ENV variables)
+- `type` (`hot`|`cold`)
+- `primary` (`true`|`false`)
+- `inserted_at`
+- `updated_at`
+
+Table name: `limits` (New)
+
+- `uuid`
+- `token_id`
+- `low_amount`
+- `high_amount`
+- `limit_type` (`percent`|`value`)
+- `blockchain_wallet_uuid`
+- `inserted_at`
+- `updated_at`
+
+In case a blockchain wallet goes below or above the limits, a user will be notified that an action is required to keep the system secure (like moving funds in/out from/to a cold wallet. In the case of the hot wallet overflowing, funds could be automatically transferred to a cold wallet.
+
+Table name: `child_chain_contract` (New) (?)
+
+**Depending on how exits are handled and how tied the watchers are with Plasma chains, this might not be needed.**
+
+- `uuid`
+- `contract_address`
+- `active` (`true`|`false`)
+- `inserted_at`
+- `updated_at`
+
+Table name: `deposit` (New) (?)
+
+- `uuid`
+- `child_chain_contract_uuid`
+- `amount`
+- `token_id`
+- `inserted_at`
+
+Table name: `exit` (New) (?)
+
+- `uuid`
+- `child_chain_contract_uuid`
+- `inserted_at`
+
+Table name: `exited_balances` (New)
+
+- `uuid`
+- `exit_uuid`
+- `amount`
+- `token_id`
+- `inserted_at`
+
+Table name: `wallet` (Existing)
+
+- `blockchain_wallet_uuid` (`nullable`)
+
+Table name: `token` (Existing)
+
+- `contract_address` (`nullable`)
+- `contract_template_uuid` (`nullable`)
+- `status` (`pending`|`confirmed`)
+
+Table name: `mint` (Existing)
+
+- `status` (`pending`|`confirmed`)
+
+Table name: `setting` (New)
+
+- `key`
+- `value`
+- `inserted_at`
+- `updated_at`
+
+Table name: `transaction` (Existing)
+
+- `type` (`internal`|`external`)
+
+If the transaction is `external`, the addresses will be from blockchain wallets.
+
+Table name: `transaction_request` (Existing)
+
+- `blockchain_wallet_address`
+
+Table name: `transaction_consumption` (Existing)
+
+- `ewallet_pub_key`
+- `blockchain_wallet_address`
+
+Keys / Values:
+
+`ewallet_identifier` : `pub_key` (`null` if not setup with blockchain)
+
+#### Flows Implementations
+
+This section describes how the flows explained earlier interact with the eWallet entities.
+
+Coming soon.
```

Comment:
Some description of the the 'unhappy paths' is required.

* Particularly the case where a user with custody of their private key sends funds to an eWallet with full custody.
* What happens if an underpayment or over payment is made?